source/server-selection/server-selection.rst

```diff
@@ -1180,11 +1185,85 @@ server without an explicit ``start_request`` (or comparable) method call.
 Outside a legacy "request" API, drivers MUST use server selection for each
 individual read operation.
 
-Reference Implementation
-========================
-
-The single-threaded reference implementation is the Perl master branch (work
-towards v1.0.0).  The multi-threaded reference implementation is TBD.
+Logging
+-------
+Please refer to the `logging specification <../logging/logging.rst>`_ for
+details on logging implementations in general, including log levels, log
+components, and structured versus unstructured logging.
+
+Drivers MUST support logging of server selection information via the
+following log messages. These messages MUST use the ``serverSelection`` log
+component.
+
+The types used in the structured message definitions below are demonstrative,
+and drivers MAY use similar types instead so long as the information is present
+(e.g. a double instead of an integer, or a string instead of an integer if the
+structured logging framework does not support numeric types.)
+
+"Waiting for suitable server to become available" message
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+This message MUST be logged at ``info`` level. It MUST be emitted on the occasions
+specified either in `Multi-threaded or asynchronous server selection`_ or
+`Single-threaded server selection`_, depending on which algorithm the driver
+implements.
+
+In order to avoid generating redundant log messages where all information besides
+the remaining time is identical, this message MUST not be repeatedly emitted for an
+operation unless the topology description changes.
+
+This message MUST contain the following key-value pairs:
+
+.. list-table::
+   :header-rows: 1
+   :widths: 1 1 1
+
+   * - Key
+     - Suggested Type
+     - Value
+
+   * - message
+     - String
+     - "Waiting for suitable server to become available"
+
+   * - selector
+     - String
+     - String representation of the selector being used to select the server. This can be
+       a read preference or an application-provided custom selector. The exact content of
+       is flexible depending on what the driver is able to log. At minimum, when the selector
+       is a read preference this string MUST contain all components of the read preference,
+       and when it is an application-provided custom selector the string MUST somehow indicate
+       that it is a custom selector.
+
+   * - requestId
```

Comment:
.Net initiates the operation info only after the server was selected, `requestId` is generated only before the message goes on the wire. But we do have `operationId` at that point.

Reply:
Adding operation information looks feasible, but it would be a significant refactoring to generate `requestId` before the server selection. And it's even more complicated with retries.
Wondering what is the complexity of this in other languages, particularly in Java.

Reply:
cc @jyemin re complexity for Java?

if it ends up being too big an undertaking, we could make the requestId a SHOULD and not a MUST, and split out the assertions for its presence into separate tests so you could skip them. for those who can do it easily enough, I think it's worthwhile to keep.

Reply:
Java has neither operationId (at all) or requestId (until much further down the stack).  In practice, requestId seems problematic anyway due to operations like bulk writes that can be split into multiple commands.  It wouldn't be that hard to add operationId though.

Reply:
>  In practice, requestId seems problematic anyway due to operations like bulk writes that can be split into multiple commands

if a driver includes both the requestId and operationId for bulk writes, how would this be problematic? it's the same thing we do for command events/log messages

Reply:
A bulk write looks like this (ignoring many details)

```
     select server
          <log server selection events> 

     while moreSplitPointsAvailable
          execute insert/update/delete command
              <log/publish events for command started>
              <log/publish events for command succeeded>
```    
     
I don't see how a driver would have a single requestId available during server selection.  

Moreover, requestId is part of the wire protocol and not something that should be available at the operation layer of abstraction.

Reply:
Jeff and I talked about this offline. It turns out at least some drivers (Java, .NET) only perform server selection once for a bulk write, and the request ID is generated all the way down at the point of creating a wire protocol message.
For Rust however, (while we don't have bulk write we do have `insert_many` at least), we re-perform server selection for each batch. So for Rust, there is a 1:1 mapping of requestId to server selection attempt, but this won't necessarily hold up for other drivers.
I can't find anything in the specs on how drivers should handle server selection for bulk writes, so I've filed DRIVERS-2502 about addressing this in the new bulk write spec.

Given all this, I don't think it makes sense to put the requestId in these messages since some (and possibly most) drivers do not have a 1:1 mapping of server selection attempt to command. That said, I will still update the messages to include the operationId, for drivers that generate those.